webapp/src/app/participant-view/participant-stage-view/exp-leader-reveal/exp-leader-reveal.component.ts

```diff
@@ -1,80 +1,66 @@
+
+    this.stageData = this.stage.config;
+    this.everyoneReachedThisStage = this.participant.everyoneReachedCurrentStage(this.stage.name);
+  }
+
+  private queryClient = injectQueryClient();
+  private mutationReveal = updateLeaderRevealStageMutation(this.queryClient, () =>
+    this.participant.navigateToNextStage(),
+  );
+
+  get stage(): ExpStageVoteReveal {
+    return this._stage as ExpStageVoteReveal;
+  }
+
+  private _stage?: ExpStageVoteReveal;
+
   public participant: Participant;
   public stageData: VoteReveal;
 
-  public everyoneReachedTheEnd: Signal<boolean>;
+  public everyoneReachedThisStage: Signal<boolean>;
   public finalLeader: Signal<string>;
 
   constructor(
     @Inject(PARTICIPANT_PROVIDER_TOKEN) participantProvider: ProviderService<Participant>,
   ) {
     this.participant = participantProvider.get();
-    const { config } = this.participant.assertViewingStageCast(StageKind.RevealVoted)!;
-
-    this.stageData = config;
+    this.stageData = this.stage?.config; // This will truly be initialized in ngOnInit. this.stage can be undefined here
 
-    // TODO: use the new backend way of syncing participant progression
-    this.everyoneReachedTheEnd = signal<boolean>(false);
-    //  computed(() => {
-    //   const users = Object.values(this.participant.experiment().participants);
-    //   const isReady = users
-    //     .map((userData) => userData.futureStageNames.length)
-    //     .every((n) => n === 1);
-    //   // Allow "Next" to be pushed.
-    //   if (isReady) {
-    //     for (const user of users) {
-    //       user.allowedStageProgressionMap[user.workingOnStageName] = true;
-    //     }
-    //   }
-    //   return isReady;
-    // });
+    this.everyoneReachedThisStage = signal<boolean>(false);
 
     // TODO: use the new backend
-    this.finalLeader = signal<string>('');
-    //  computed(() => {
-    //   const users = Object.values(this.participant.experiment().participants);
-    //   const votes: { [userId: string]: number } = {};
-    //   users.forEach(({ uid }) => {
-    //     votes[uid] = 0;
-    //   });
-
-    //   for (const user of users) {
-    //     const leaderVotes = user.stageMap[this.stageData.pendingVoteStageName].config as Votes;
-    //     for (const userId of Object.keys(leaderVotes)) {
-    //       const vote = leaderVotes[userId];
-    //       if (vote === 'positive') {
-    //         votes[userId] += 1;
-    //       } else if (vote === 'negative') {
-    //         votes[userId] -= 1;
-    //       } else if (vote === 'neutral') {
-    //         votes[userId] += 0;
-    //       }
-    //     }
-    //   }
-
-    //   const sorted = reverse(
-    //     sortBy(
-    //       Object.entries(votes).map(([userId, vote]) => ({ userId, vote })),
-    //       ['vote'],
-    //     ),
-    //   );
+    this.finalLeader = signal<string>('TODO');
+  }
 
-    //   return sorted[0].userId;
-    // });
+  nextStep() {
+    this.mutationReveal.mutate({
```

Comment:
I think we need a comment to say what/why we are doing this...

Reply:
yeah, the idea was to structure the stage flow like so:
1. You fill the stage data as a participant
2. You use a mutation to send your data to the backend / firestore
3. If the mutation was successful, you navigate to the next page

However, I realized later on that participants are not supposed to be able to edit previous stages, so we should be able to simplify this (if the stage is being worked on by the participant, mutation -> navigation. Else, just navigate).

Because the LeaderReveal stage has no data to be filled by the participant, this mutation feels especially heavy (all that work just to update `workingOnStageName`), I will think of lighter alternatives

Reply:
oh I think I see, the point here is that you save the data to firestore at the time the person does the nextStep thing. Previously I did writes when data was edited... might be worth discussing how hard this is, and what the right abstraction is. My default would be that each stage kind has a class, and that class has functions to edit the data, and when edited, that results in a write to Firebase. So the key thing in a nextStep would be to call the corresponding stage instance's nextStep method that would simply mark this stage as complete.